src/rust/protocol/src/memcache/wire/request/parse.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use super::super::*;
+use crate::*;
+
+use core::slice::Windows;
+use std::convert::TryFrom;
+
+const MAX_BYTES: usize = usize::MAX / 2;
+
+impl Parse for MemcacheRequest {
+    fn parse(buffer: &[u8]) -> Result<ParseOk<Self>, ParseError> {
+        match parse_command(buffer)? {
+            MemcacheCommand::Get => parse_get(buffer),
+            MemcacheCommand::Gets => parse_gets(buffer),
+            MemcacheCommand::Set => parse_set(buffer, false),
+            MemcacheCommand::Add => parse_add(buffer),
+            MemcacheCommand::Replace => parse_replace(buffer),
+            MemcacheCommand::Cas => parse_set(buffer, true),
+            MemcacheCommand::Delete => parse_delete(buffer),
+        }
+    }
+}
+
+struct ParseState<'a> {
+    single_byte: Windows<'a, u8>,
+    double_byte: Windows<'a, u8>,
+}
+
+impl<'a> ParseState<'a> {
+    fn new(buffer: &'a [u8]) -> Self {
+        let single_byte = buffer.windows(1);
+        let double_byte = buffer.windows(2);
+        Self {
+            single_byte,
+            double_byte,
+        }
+    }
+
+    fn next_space(&mut self) -> Option<usize> {
+        self.single_byte.position(|w| w == b" ")
+    }
+
+    fn next_crlf(&mut self) -> Option<usize> {
+        self.double_byte.position(|w| w == CRLF.as_bytes())
+    }
+}
+
+fn parse_command(buffer: &[u8]) -> Result<MemcacheCommand, ParseError> {
+    let command;
+    {
+        let mut parse_state = ParseState::new(buffer);
+        if let Some(_line_end) = parse_state.next_crlf() {
```

Comment:
This is vulnerable to invalid commands that doesn't terminate with a space or `CRLF` within the first X bytes (X being something like 10) as the iterator will keep moving till we exhaust the byte buffer. Is there an iterator method that allows us to bound the distance of the first space? Or can you re-slice the buffer to achieve that?

You can make this a TODO but I do think it makes it easier to (accidentally) DoS the service via parser behavior if users send garbage data.

Reply:
It should be harder to DoS now. Added batch size, command length, and key length limits. Some of the early parsing could probably be improved.